core/src/connection_reuse.rs

```diff
@@ -40,193 +40,543 @@
 //! `MuxedTransport` trait.
 
 use fnv::FnvHashMap;
-use futures::future::{self, Either, FutureResult};
-use futures::{Async, Future, Poll, Stream};
+use futures::future::{self, FutureResult};
 use futures::stream::FuturesUnordered;
-use futures::sync::mpsc;
+use futures::{stream, task, Async, Future, Poll, Stream};
 use multiaddr::Multiaddr;
 use muxing::StreamMuxer;
 use parking_lot::Mutex;
-use std::io::{self, Error as IoError};
-use std::sync::Arc;
+use std::collections::{HashMap, hash_map::Entry};
+use std::io::{Error as IoError, ErrorKind as IoErrorKind, Read, Write};
+use std::ops::{Deref, DerefMut};
+use std::sync::{atomic::AtomicUsize, atomic::AtomicBool, atomic::Ordering, Arc};
 use tokio_io::{AsyncRead, AsyncWrite};
 use transport::{MuxedTransport, Transport, UpgradedNode};
 use upgrade::ConnectionUpgrade;
 
+use std::hash::{BuildHasherDefault, Hasher};
+
+#[derive(Default)]
+struct UsizeHasher {
```

Comment:
Why not use `FnvHashMap`, which afaik does this better, instead of defining our own hasher?

Reply:
Performance. [See the conversation I had with @twittner about this](https://github.com/libp2p/rust-libp2p/pull/387#discussion_r208851171). I am not sure what you mean with `FnvHashMap` doing "this better", `FnvHasher` does the same we do here: [it defines its own hasher](https://doc.servo.org/src/fnv/lib.rs.html#114-121), using the Fnv-Algo. However as we already have pretty unique values, that Algo is a bit more expensive (~10-15% in this case) than us just converting the `usize`-Key into the required `u64` type. We are doing the exact same as `Fnv` but with a less expensive "algorithm" (if you can call it that). 

This algorithm however is very much constraint to the use case at hand and should probably not be re-used at many other places. Which is why I can understand the argument to go for a more generic solution - if we are willing to pay the performance price, I am fine with doing that, too.